### DISCUSSION TITLE
[console] why signal handlers on SignableCommandInterface exit immediately after last handler

### DISCUSSION THREAD

**Discussion by akuzia** [symfony/symfony#48024] [Q&A] [answered]
I've tried to use SignableCommandInterface to control long running tasks and stumbled upon this: as seen in [Application](https://github.com/symfony/console/blob/6.1/Application.php#L1005) process exits after last handler. What "default behavior" was mentioned above? Process simply dies after running all handlers and `execute` is never finished. What is the motivation behind this `exit`?

So scenarios where there is a controlled loop in execute and some kind of a breaker in signal handlers are not achievable with `SignableCommandInterface` and must be implemented outside of interface or has I misunderstood smth?

What is the intention behind this interface? Simple cleanup?

Behavior I consider default:
```php
<?php

class foo
{
    private bool $running = true;

    public function __construct()
    {
        pcntl_async_signals(true);

        pcntl_signal(\SIGINT, function (): void {
            $this->shutdown();
        });
    }

    public function shutdown(): void
    {
        echo 'Terminating...' . PHP_EOL;
        $this->running = false;
    }

    public function run(): void
    {
        while ($this->running) {
            echo 'Running...' . PHP_EOL;
            sleep(1);
        }
        echo 'Exited' . PHP_EOL;
    }
}

$runner = new foo();
$runner->run();
```

```
php foo.php 
Running...
Running...
^CTerminating...
Exited
```

**Comment by MatTheCat** [comment]
> Process simply dies after running all handlers and `execute` is never finished. What is the motivation behind this `exit`?

The process only dies if this is what is expected from the signal: https://github.com/symfony/symfony/blob/541e651560ed3fce7a9c512a94c7aff8bfb2fc4f/src/Symfony/Component/Console/Application.php#L1015-L1017

> So scenarios where there is a controlled loop in execute and some kind of a breaker in signal handlers are not achievable with `SignableCommandInterface` and must be implemented outside of interface or has I misunderstood smth?

Your command does not need to implement `SignableCommandInterface` if you just want it to exit on `SIGINT` or `SIGTERM`. As you can see in the code above, this is the default behavior.

> What is the intention behind this interface? Simple cleanup?

Whatever you want to do when receiving a signal. As mentioned in [the original blog post](https://symfony.com/blog/new-in-symfony-5-2-console-signals):

> (e.g. to perform some cleanup)

**Comment by akuzia** [comment, marked as answer]
> The process only dies if this is what is expected from the signal

SIGTERM does not expect process to exit immediately, it [can be ignored](https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html#index-SIGTERM). It is a "polite" way to ask program to exit.

So:

```php
<?php

namespace App\Command;

use Symfony\Component\Console\Attribute\AsCommand;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Command\SignalableCommandInterface;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;

#[AsCommand(
    name: 'run'
)]
class TestCommand extends Command implements SignalableCommandInterface
{
    private bool $running = true;

    public function getSubscribedSignals(): array
    {
        return [\SIGINT];
    }

    public function handleSignal(int $signal): void
    {
        echo sprintf("Recieved signal %s\n", $signal);
        $this->running = false;
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $io = new SymfonyStyle($input, $output);
        while ($this->running) {
            $io->writeln('running...');

            sleep(1);
        }

        $io->success('You have exited after signal!');

        return Command::SUCCESS;
    }
}
```
results in: 
```
./bin/console run                                                                                                                               ✔ 
running...
running...
^CRecieved signal 2
```

And:

```php
<?php

namespace App\Command;

use Symfony\Component\Console\Attribute\AsCommand;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Command\SignalableCommandInterface;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;

#[AsCommand(
    name: 'run'
)]
class TestCommand extends Command
{
    private bool $running = true;

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        pcntl_async_signals(true);

        pcntl_signal(\SIGINT, function (): void {
            echo sprintf("Recieved signal");
            $this->running = false;
        });
        $io = new SymfonyStyle($input, $output);
        while ($this->running) {
            $io->writeln('running...');

            sleep(1);
        }

        $io->success('You have exited after signal!');

        return Command::SUCCESS;
    }
}

```
is ok:

```
./bin/console run                                                                                                                           255 ✘ 
running...
running...
^CRecieved signal
                                                                                                                        
 [OK] You have exited after signal!                                                                                     
                                                                                                                        
```

:man_shrugging:

**Reply by MatTheCat** [reply]
You don’t even need to bother with signal handling in these commands as Symfony will do it for you (I don’t understand the point you’re trying to make?)

Maybe you could open a PR to allow not reacting on SIGTERM but I feel this is not really a problem?

(Also please write in “Write a reply” instead of “Suggest an answer” to reply to a message.)

**Reply by MatTheCat** [reply]
Oh I just saw https://github.com/symfony/symfony/issues/47809 is that what you meant?

**Reply by akuzia** [reply]
Yep, this is it

**Reply by MatTheCat** [reply]
Okay, well, at least now you have an issue to subscribe to!

**Reply by lyrixx** [reply]
I'm marking this discussion as solved, since there is an issue for it. Let's discuss it there. 

Thanks for trying to enhance Symfony :+1:
